### PR TITLE
SageMaker Training controller with e2e test

### PR DESCRIPTION
**Pull request by mbaijal**
## Description of changes:
Controller for SageMaker TrainingJob resource with integration test. 
 - Updated generator.yaml config file with a specific operation for `StopTrainingJob` and various Status fields
 - code-generated CRD, controller, rbac and manifests to help create installers.
 - e2e tests to create and stop a trainingjob. 
 - Relevant dataset has been added to the bucket.
 - I have also made a minor edit to the `kind-build-test.sh` to print the debug message at the end of the test runs. 

## Pending:
 - Add a trainingJob Test for debugger
 - Add another create test which waits for the job to reach completion
 - Enable SageMaker test scripts to handle multiple test runs of the same kind more seamlessly (Name generation, more modular scripts etc)
 - Find a smaller dataset, or possibly generate one such removing the dependency on the source buckets. 
 - the format of the debug message is not in sync with the rest of the repo. The `[ERROR]` messages in the SageMaker scripts make debugging easier. 
 - Within SageMaker scripts, the docstrings do not follow the same template. 

## Testing
Using personal account, followed instructions on the ACK testing page. 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: scripts/lib/aws/sagemaker.sh [hunk 1]
```diff
@@ -65,12 +66,12 @@ sagemaker_create_execution_role() {
 
   local __role_name="$1"
   daws iam create-role --role-name "$__role_name" --assume-role-policy-document '{"Version": "2012-10-17","Statement": [{ "Effect": "Allow", "Principal": {"Service": "sagemaker.amazonaws.com"}, "Action": "sts:AssumeRole"}]}' >/dev/null
-  assert_equal "0" "$?" "Expected success from aws iam create-role --role-name $__role_name but got $?" || exit 1
+  assert_equal "0" "$?" "Expected success from aws iam create-role --role-name $__role_name, got $?" || exit 1
```
Comment (jaypipes) [inline scripts/lib/aws/sagemaker.sh:69]: I mean, it's up to you, but any reason to change these assertion failure messages? Oftentimes I like to keep style-only changes like this in a separate PR that can reviewed apart from functional changes.

File: scripts/lib/aws/sagemaker.sh [hunk 2]
```diff
@@ -195,9 +197,150 @@ assert_aws_sagemaker_model_arn() {
...
+# Arguments:
+#
+#   TRAINING_JOB_NAME  required string for the name of the bucket to check
+#
+# Usage:
+#
+#   if ! training_job_created "$training_job_name"; then
+#       echo "Training Job $training_job_name does not exist!"
+#   fi
+is_aws_sagemaker_trainingjob_exists() {
```
Comment (jaypipes) [inline scripts/lib/aws/sagemaker.sh:220]: I think "aws" is redundant :) Also, to match functions in other `scripts/lib/aws` modules, might be good to call this `sagemaker_trainingjob_exists()`. For example, see the [`sns_topic_exists()` function](https://github.com/aws/aws-controllers-k8s/blob/cc3c433f3407df06554cd7d937f7e0286b33bc59/scripts/lib/aws/sns.sh#L26).

File: scripts/lib/aws/sagemaker.sh [hunk 3]
```diff
@@ -195,9 +197,150 @@ assert_aws_sagemaker_model_arn() {
...
+#
+# get_aws_sagemaker_trainingjob_status TRAINING_JOB_NAME
+# Arguments:
+#
+#   TRAINING_JOB_NAME  required string for the name of the trainingJob to check
+#
+# Usage:
+#
+#   local training_job_status=$(get_aws_sagemaker_trainingjob_status "${__training_job_name}")
+get_aws_sagemaker_trainingjob_status() {
```
Comment (jaypipes) [inline scripts/lib/aws/sagemaker.sh:274]: `sagemaker_training_job_get_status()`?

File: scripts/lib/aws/sagemaker.sh [hunk 4]
```diff
@@ -195,9 +197,150 @@ assert_aws_sagemaker_model_arn() {
...
+  if [[ "$#" -lt 1 || -z ${AWS_REGION+x} ]]; then
+    echo "[FAIL] Usage: is_aws_sagemaker_trainingjob_exists training_job_name. Expects AWS_REGION to be set"
+    exit 1
+  fi
+  local __training_job_name="$1"
+
+  local __training_job_status="$(daws sagemaker describe-training-job --region "$AWS_REGION" --training-job-name $__training_job_name --output json | jq .TrainingJobStatus)"
+
+  if [[ $? -eq 0 ]]; then
+      echo "$__training_job_name found!"
```
Comment (jaypipes) [inline scripts/lib/aws/sagemaker.sh:230]: recommend not placing `echo` statements in functions that just return a 0 or 1. Instead, place `echo` statements where you *call* these functions, if at all.

File: test/e2e/sagemaker/e2e.sh [hunk 5]
```diff
@@ -15,8 +15,19 @@ source "$SCRIPTS_DIR/lib/aws/sagemaker.sh"
 
 sagemaker_setup_common_test_resources
 
+# Test TrainingJob
+# TODO: A cleaner way to do this
+THIS_DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null 2>&1 && pwd )"
```
Comment (jaypipes) [inline test/e2e/sagemaker/e2e.sh:20]: THIS_DIR is defined below on line 29. Can you remove line 29 please?

File: test/e2e/sagemaker/helper/trainingjob/xgboost-train.sh [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+  if [[ -z ${RANDOM_ID+x} || -z ${AWS_REGION+x} || -z ${SAGEMAKER_DATA_BUCKET+x} ]]; then
+    echo "[FAIL] Usage setup_sagemaker_training_test_inputs. Expects RANDOM_ID, SAGEMAKER_DATA_BUCKET and AWS_REGION to be set"
+    exit 1
+  fi
+
+  TRAINING_JOB_NAME="ack-trainingjob-${RANDOM_ID}"
+  # TrainingJobName must have length less than 64
+  TRAINING_JOB_NAME=`echo $TRAINING_JOB_NAME|cut -c1-64`
+  TRAINING_IMAGE_REGISTRY=`echo $(get_xgboost_registry $AWS_REGION)`
+  TRAINING_IMAGE_USWEST2="$TRAINING_IMAGE_REGISTRY.dkr.ecr.$AWS_REGION.amazonaws.com/xgboost:1"
```
Comment (jaypipes) [inline test/e2e/sagemaker/helper/trainingjob/xgboost-train.sh:31]: nothing for this particular PR, but generally I shy away from using script-scoped or global-scoped variables like this inside a function. Better to just set these variables outside of a function in the test that is actually setting this up :)

File: test/e2e/sagemaker/helper/trainingjob/xgboost-train.sh [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+# PRE-CHECKS
+#################################################
+sagemaker_trainingjob_prechecks() {
+    setup_sagemaker_training_test_inputs
+    if !is_aws_sagemaker_trainingjob_exists "$TRAINING_JOB_NAME"; then
+        echo "[FAIL] expected $TRAINING_JOB_NAME to not exist in SageMaker. Did previous test run cleanup?"
+        exit 1
+    fi
+
+    if k8s_resource_exists "$TRAINING_JOB_NAME"; then
```
Comment (surajkota) [inline test/e2e/sagemaker/helper/trainingjob/xgboost-train.sh:96]: +1

File: test/e2e/sagemaker/helper/trainingjob/xgboost-train.sh [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+verify_trainingjob_created() {
+  if [[ "$#" -lt 1 ]]; then
+    echo "[FAIL] Usage: verify_trainingjob_created training_job_name"
+    exit 1
+  fi
+
+  local __training_job_name="$1"
+  # Verify TrainingJob ARN was populated in Status
+  local __k8s_trainingjob_arn=$(get_field_from_status "trainingjob/$__training_job_name" "ackResourceMetadata.arn")
+  # Verify model exists on SageMaker and matches ARN populated in Status
```
Comment (surajkota) [inline test/e2e/sagemaker/helper/trainingjob/xgboost-train.sh:117]: correct the comment

File: test/e2e/sagemaker/helper/trainingjob/xgboost-train.sh [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+    assert_equal "0" "$?" "FAILED: Expected success from kubectl delete but got $?" || exit 1
+    assert_aws_sagemaker_trainingjob_stopped "$TRAINING_JOB_NAME" || exit 1
+}
+
+#################################################
+# Post-Checks
+#################################################
+sagemaker_trainingjob_postchecks() {
+    debug_msg "Check that the controller pod was not restarted."
+    assert_pod_not_restarted $ack_ctrl_pod_id
```
Comment (surajkota) [inline test/e2e/sagemaker/helper/trainingjob/xgboost-train.sh:147]: this should be a common postcheck for all tests ?

File: services/sagemaker/cmd/controller/main.go [hunk 10]
```diff
@@ -28,6 +28,7 @@ import (
 	svcresource "github.com/aws/aws-controllers-k8s/services/sagemaker/pkg/resource"
 
 	_ "github.com/aws/aws-controllers-k8s/services/sagemaker/pkg/resource/model"
+	_ "github.com/aws/aws-controllers-k8s/services/sagemaker/pkg/resource/training_job"
```
Comment (surajkota) [inline services/sagemaker/cmd/controller/main.go:31]: @jaypipes why do we have these imports

**Review by surajkota — CHANGES_REQUESTED** [review]

File: test/e2e/sagemaker/e2e.sh (hunk 5 above)
Comment (jaypipes) [reply test/e2e/sagemaker/e2e.sh:20]: Don't use `source` then. Instead use `.`

File: services/sagemaker/generator.yaml [hunk 11]
```diff
@@ -1,4 +1,21 @@
...
+        404: ValidationException
+    status_fields:
+    - operation_id: DescribeTrainingJob
+      source_name: TrainingJobStatus
+    - operation_id: DescribeTrainingJob
+      source_name: SecondaryStatus
+    - operation_id: DescribeTrainingJob
+      source_name: FailureReason
+    - operation_id: DescribeTrainingJob
+      source_name: DebugRuleEvaluationStatuses
```
[the remark this answers is not in the document]
Comment (mbaijal) [reply services/sagemaker/generator.yaml:18]: You I had assumed the same, but not the case - the trainingJobStatus gets updated to the correct completed status without the requeue change.

[the remark this answers is not in the document]
Comment (mbaijal) [reply services/sagemaker/generator.yaml:18]: ++ about the debug status, need to look into that.

**Comment by mbaijal** [comment]
Closing this Pull Request since the SageMaker Tests are now being written using Python.

**Pull request closed without merging**